DSDA Lua script

kalimag opened the pull request:
(continued from #4458)

- Split definitions out into reusable module
- Add sector and line definitions

TODO:
- [x] Complete player and mobj structs
- [x] Heretic and Hexen enums
- [x] ~~Misc domain?~~
- [x] Make structs more convenient to use
  Ideally you can write code like `for mobj in pairs(dsda.mobjs) do print(mobj.type) end` and all the memory reading is abstracted away
  - [x] Better array handling
- [x] Remove `Objects` cache in `things-lines.lua` (doesn't work with spawned objects)
- [x] Cache lines instead?
  Seems to be a big performance cost on fancy Hexen maps
- [ ] Documentation
- [ ] Example code

If time permits:
- [x] Center zoom
- [x] Filter more mobj types (blood, smoke, decorations, etc.)
- [x] Stop shooting while clicking the map buttons
  - ~~Annoyingly, moving the map while paused isn't possible because of the way `gui.text` works~~

Check if completed:
- [x] I, the commit author, have read the [licensing terms for contributors](https://github.com/TASEmulators/BizHawk/blob/master/contributing.md#copyrights-and-licensing) (last updated 2024-06-22) and am compliant

kalimag committed:
Split definitions into reusable module

kalimag committed:
Add 0 values to enums

kalimag committed:
Make `struct.align` chainable

kalimag committed:
Add `u16` and `float`

kalimag committed:
Add sector and line definitions

kalimag committed:
Add links to source

kalimag committed:
Highlight special lines

kalimag committed:
Complete player and mobj defs

kalimag committed:
Spacing

vadosnaprimer commented on Assets/Lua/Doom/dsda-data.lua:
```diff
@@ -156,6 +157,126 @@ dsda.mobj = dsda.struct_layout()
+	.s16  ("tag")
+	.s32  ("cachedheight")
+	.s32  ("scaleindex")
+  	-- e6y
```
That's just signature of who added it to the code.
https://doomwiki.org/wiki/Andrey_Budko_(entryway)

vadosnaprimer commented on Assets/Lua/Doom/dsda-data.lua:
```diff
@@ -156,6 +157,126 @@ dsda.mobj = dsda.struct_layout()
+	.s16  ("colormap")
+	.ptr  ("touching_thinglist")
+	.s32  ("linecount")
+	.ptr  ("lines")
```
XDRE shows these when you enter sector ID
<img width="155" height="134" alt="изображение" src="https://github.com/user-attachments/assets/98727f43-fb5c-43ea-845f-75348a19a9c1" />

kalimag committed:
Add Heretic and Hexen enum values

kalimag committed:
Enhance struct helper to encapsulate memory reading

kalimag committed:
Use dsda-data to iterate

kalimag committed:
Use dsda-data to replace memory reads

kalimag committed:
Suppress firing while clicking on map buttons

kalimag committed:
Clear drawings when script is disabled

vadosnaprimer committed:
fix maybe_swap() to return a tuple instead of doing nothing

kalimag committed:
Stop caching mobjs

kalimag committed:
Cache lines

vadosnaprimer committed:
work while paused

vadosnaprimer committed:
center out zooming

vadosnaprimer committed:
wheel zoom!
cleanup

vadosnaprimer committed:
mouse dragging (requires shift key)

vadosnaprimer committed:
account for merged modifiers

kalimag committed:
Add mobjflags enum

kalimag committed:
Color mobjs by type

Also reorder code to minimize reads for mobjs that aren't displayed

kalimag committed:
Fix corpse color

kalimag committed:
Rename u64 -to s64

Lua doesn't have unsigned longs, but the signedness doesn't matter for flags and pointers anyway

kalimag committed:
Local mem functions

kalimag committed:
Track overall struct alignment

kalimag committed:
Support reading structs from system bus

kalimag committed:
Add debug flag for pointer correctness

kalimag committed:
Read line vertices from system bus

kalimag committed:
Fix reading from the wrong domain

kalimag committed:
Return nil for null pointers

kalimag committed:
More typed pointers

vadosnaprimer commented on Assets/Lua/Doom/dsda-data.lua:
```diff
@@ -0,0 +1,1864 @@
+dsda.LINE_SIZE   = 256  -- sizeof(line_t) is 232, but we padded it for niceness
+dsda.MOBJ_SIZE   = 512  -- sizeof(mobj_t) is 464, but we padded it for niceness
+dsda.PLAYER_SIZE = 1024 -- sizeof(player_t) is 729, but we padded it for niceness
+dsda.SECTOR_SIZE = 512  -- sizeof(sector_t) is 344, but we padded it for niceness
```
Maybe a `Sizes` table for these too?

vadosnaprimer commented on Assets/Lua/Doom/dsda-data.lua:
```diff
@@ -0,0 +1,1864 @@
+local read_u8   = memory.read_u8
+local read_u24  = memory.read_u24_le
+local read_u32  = memory.read_u32_le
+local readfloat = memory.readfloat
```
Similar shortcuts exist in the main script, looks like it makes sense to declare them once and then use everywhere?

kalimag replied:
I don't think it makes sense to share them across scripts, but in any case they're not used anymore in `things-lines.lua` because all the memory reads happen through the dsda script now. 

Unless you would like to keep them for something, I'll just remove them from `things-lines.lua`

vadosnaprimer replied:
Okay, please remove them then!

kalimag committed:
Pad structs, verify expected size

kalimag committed:
Remove size consts

They're still available through `line.padded_size` etc.

kalimag committed:
Add mobjinfo struct

kalimag committed:
Another typed pointer, comments

kalimag committed:
Fix pointer check for null pointers

kalimag committed:
More lenient 32bit/64bit pointer handling

kalimag committed:
Support adding "properties" and functions to struct instances

kalimag committed:
Add state and msecnode structs

kalimag committed:
Add functions for iterating linked lists

kalimag committed:
Add subsector struct

kalimag committed:
Color table, minimum zoom levels for mobjs

kalimag committed:
Add more categories, filter some mobjs by type

kalimag committed:
More caching, workaround for double execution

kalimag committed:
Remove unused locals

kalimag committed:
Add back size tables

kalimag committed:
Add struct names

kalimag committed:
Support nested structs

kalimag committed:
Turn sector.damage into nested struct

kalimag committed:
Separate struct building from struct

kalimag committed:
Split up file

kalimag committed:
Expose array fields as tables

kalimag committed:
Add utility functions for arrays, wrap `sector.lines`

kalimag committed:
Add `inventory_t `

kalimag committed:
Bounding boxes as array

kalimag committed:
Rename parameters

kalimag committed:
Add array offset

kalimag committed:
Add symbol table

kalimag committed:
Show idle enemies in a different color

kalimag committed:
Supress click inputs when interacting with map

kalimag committed:
Split method for structs with/without dedicated domain

kalimag committed:
Change cache init

kalimag committed:
Expose variables directly from system bus

kalimag committed:
Update symbols

kalimag committed:
Bypass Things domain

kalimag committed:
Handle float array indices

kalimag committed:
`global_layout` changes

kalimag committed:
Expose more global arrays

kalimag committed:
Automatically handle null pointers to arrays

kalimag committed:
Add functions for reading line coords efficiently

kalimag committed:
Clear map cache on map changes

kalimag committed:
Overhaul line caching

Now we can tell which lines are actually polyobjs/scriptable

Also read lines from system bus

kalimag committed:
Fix nil error in vanilla Doom

vadosnaprimer commented on Assets/Lua/Doom/things-lines.lua:
```diff
@@ -267,7 +267,7 @@ end
 	if Init then return end
 
-	for addr, mobj in pairs(structs.mobj.items) do
+	for _, mobj in pairs(structs.global.mobjs) do
```
I should implement reading other things than 1 byte...

vadosnaprimer replied:
Fair! And with array sizes changing, only Players can be turned into a pointer based domain, but players are far from majority of this script so I don't think it'd matter.

kalimag committed:
Update symbols

kalimag committed:
Read players from system bus

kalimag committed:
Use bulk reads for pointer arrays

kalimag committed:
Reuse screen dimensions

kalimag committed:
Remove unused declarations, misc cleanup

kalimag committed:
Implement struct equality, handle circular linked lists

kalimag committed:
More globals, reorder stuff

kalimag committed:
Rename globals

kalimag committed:
Unify from_* methods a bit

kalimag committed:
Struct array shortcuts

kalimag committed:
Move asserts

kalimag committed:
Misc cleanup

kalimag committed:
Handle non-numeric/non-sequential enum tables

kalimag committed:
Add a few enums

kalimag committed:
Add some annotations

Help LLS find modules

kalimag committed:
Support symbol lookup by address

kalimag committed:
Add spacing
